src/common/thread.pool.cpp

```diff
@@ -0,0 +1,155 @@
+#include "thread.pool.hh"
+#include "utilities.hh"
+
+namespace zarr = acquire::sink::zarr;
+namespace common = zarr::common;
+
+common::ThreadPool::ThreadPool(size_t n_threads,
+                               std::function<void(const std::string&)> err)
+  : error_handler_{ err }
+  , is_accepting_jobs_{ true }
+{
+    n_threads = std::clamp(
+      n_threads,
+      (size_t)1,
+      (size_t)std::max(std::thread::hardware_concurrency(), (unsigned)1));
+
+    for (auto i = 0; i < n_threads; ++i) {
+        threads_.emplace_back([this] { thread_worker_(); });
+    }
+}
+common::ThreadPool::~ThreadPool() noexcept
+{
+    {
+        std::scoped_lock lock(jobs_mutex_);
+        while (!jobs_.empty()) {
+            jobs_.pop();
+        }
+    }
+
+    await_stop();
+}
+void
+common::ThreadPool::push_to_job_queue(JobT&& job)
```

Comment:
I think that `lock.unlock()` is called unnecessarily before notifying `cv_`, which can lead to a race condition if another thread is waiting on the condition variable.

```cpp
{
    std::unique_lock lock(jobs_mutex_);
    CHECK(is_accepting_jobs_);
    
    jobs_.push(std::move(job));
}

cv_.notify_one();
```

Reply:
You may be right, but it's always been tough for me to reason about these things. My understanding of what's going on at this moment is that all threads are either going to be executing a job or waiting on the condition variable. Those threads that are executing will then cycle back to the beginning of the loop and try to acquire a lock on the mutex, which they immediately relinquish when they start waiting on the condition variable. The potential race condition I see here is that

1. the thread pool unlocks the lock
2. a worker thread acquires the lock
3. the thread pool calls `cv_.notify_one()`
4. the worker starts waiting on the condition variable after (3).

This should be fine, right? Some other thread can pick up the notify. This is of course assuming there's more than one worker thread.

Reply:
It sounds about right, but I think it’s based on assumptions that may not always hold true. When you notify the condition variable after releasing the lock, you risk a situation where:

1. The thread pool unlocks the mutex.
2. A worker thread acquires the lock, but it doesn't see the new job yet because it hasn’t started waiting on the condition variable.
3. The thread pool calls `cv_.notify_one()`.
4. The worker thread then starts waiting on the condition variable after the notify, missing the notification.

I think this can lead to inefficiencies where the notified thread doesn't see the job that was just pushed, causing it to go back to waiting unnecessarily, while another worker thread might pick up the job eventually. It's generally best practice to hold the lock while notifying to ensure that the notified thread immediately sees the state change.